### PR TITLE
Unify alignment constants between the GC and the VM and assert that t…

### PR DESCRIPTION
**Pull request by swgillespie**
…hey are equivalent. This is the second set of changes discussed in #7501. The `DATA_ALIGNMENT` macro is defined in `src/vm/<arch>/cgencpu.h`, so in `gcenv.ee.cpp` it is asserted that this macro and the one exported by the GC are the same.

For `amd64`, where we can allocate in assembly helpers, constants are introduced for the alignment constants (`7` and `~7`) and those are asserted against what the GC uses as well. In all but two cases, the requested allocation size is already aligned - for those two cases I used the introduced constants instead of immediate values.

### CODE REVIEW

**Comment by swgillespie** [comment]
cc @jkotas @Maoni0 @sergiy-k @adityamandaleeka

File: src/gc/gcpriv.h [hunk 1]
```diff
@@ -871,8 +871,7 @@ struct seg_mapping
 #endif //SEG_MAPPING_TABLE
 
 // alignment helpers
-//Alignment constant for allocation
-#define ALIGNCONST (DATA_ALIGNMENT-1)
+#define ALIGNCONST GC_ALIGNCONST
```
Comment (jkotas) [inline src/gc/gcpriv.h]: Why not find&replace ALIGNCOST with GC_ALIGNCONST in the whole GC codebase?

File: src/gc/env/gcenv.base.h [hunk 2]
```diff
@@ -293,7 +293,7 @@ typedef DPTR(uint8_t)   PTR_uint8_t;
 
 // -----------------------------------------------------------------------------------------------------------
 
-#define DATA_ALIGNMENT sizeof(uintptr_t)
+#define DATA_ALIGNMENT GC_DATA_ALIGNMENT
```
Comment (jkotas) [inline src/gc/env/gcenv.base.h]: Find&replace DATA_ALIGNMENT with GC_ALIGNCONST in the whole GC codebase?

File: src/gc/gcpriv.h (hunk 1 above)
Comment (swgillespie) [inline src/gc/gcpriv.h]: will do!

File: src/gc/env/gcenv.base.h (hunk 2 above)
Comment (swgillespie) [inline src/gc/env/gcenv.base.h]: will do!

File: src/vm/amd64/JitHelpers_InlineGetThread.asm [hunk 3]
```diff
@@ -175,8 +175,8 @@ LEAF_ENTRY AllocateStringFastMP_InlineGetThread, _TEXT
         ; Calculate the final size to allocate.
         ; We need to calculate baseSize + cnt*2, then round that up by adding 7 and anding ~7.
```
Comment (adityamandaleeka) [inline src/vm/amd64/JitHelpers_InlineGetThread.asm]: Update the comments too (they still refer to 7/8)?

**Comment by Maoni0** [comment]
>         lea     r8d, [r8d + ecx*2 + 7]

this also needs to be changed

---
Refers to: src/vm/amd64/JitHelpers_Slow.asm:602 in 9a096a2. [](commit_id = 9a096a2cb0493560c30d4dc65f20f3ccdca9ba85, deletion_comment = False)

**Comment by Maoni0** [comment]
>         add     r8d, 7

this also needs to be changed to use the consts you defined

---
Refers to: src/vm/amd64/JitHelpers_Slow.asm:668 in 9a096a2. [](commit_id = 9a096a2cb0493560c30d4dc65f20f3ccdca9ba85, deletion_comment = False)

**Comment by Maoni0** [comment]
> #define ASM_LARGE_OBJECT_SIZE 85000

can you also assert this to be the same as LARGE_OBJECT_SIZE you defined in gcinterface.h?

---
Refers to: src/vm/amd64/asmconstants.h:611 in 9a096a2. [](commit_id = 9a096a2cb0493560c30d4dc65f20f3ccdca9ba85, deletion_comment = False)

**Comment by jkotas** [comment]
> I also don't see any changes for other platforms...the sizes need to be aligned on those as well

The changes are not complete even on amd64, e.g. there are more places in jithelpers_slow.asm where the alignment is hardcoded as 7 or -8. Also, the alignment is often hardcoded in indirect way, via bit count - https://github.com/dotnet/coreclr/blob/master/src/vm/arm/patchedcode.asm#L295.

There is like hundred places where the alignment is hardcoded. I do not think it is worth spending time on chasing them down and fixing all of them at this point. You won't be able to chase all of them down anyway unless you spent time testing with different alignments on all platforms (a lot of work); and you probably introduce some regressions along the way as well.

**Comment by Maoni0** [comment]
yeah, I pointed out the ones in jithelpers_slow.asm above. does that still not make the changes for amd64 complete? If not, where are the other places in amd64? I saw the ones in arm.

we are trying to cover when we are aligning in the allocation paths. we do have quite a few allocation helpers but hundreds seem excessive.

if there are indeed hundreds, then we might as well not change it at all. it seems pointless to change only a few out of "hundreds".

**Comment by jkotas** [comment]
> we are trying to cover when we are aligning in the allocation paths

The alignment is not limited to allocation paths. A few examples: 

- SOS needs this to walk the heap https://github.com/dotnet/coreclr/blob/master/src/ToolBox/SOS/Strike/strike.h#L117

- Type builder needs it to align base size: https://github.com/dotnet/coreclr/blob/master/src/vm/methodtablebuilder.cpp#L10919 Notice that it uses `ALLOC_ALIGN_CONSTANT` - different from the constant that you are using here.

**Comment by Maoni0** [comment]
note that the purpose of this change is not for general alignment. it's for making sure the VM side is laying out objects with the same alignment that GC uses to walk the heap, ie GC gets the size of an object and align it in order to advance to the next object. so I care about the allocation paths as they are the ones that actually change the heap. good point about sos though - I forgot about it. it also walks the heap but there shouldn't be many places that are affected in sos. 

we looked at the base size alignment - since GC gets the base size from the methodtable, it doesn't affect how GC walks the heap, as GC gets the base size from the VM. 

do you still think there are hundreds of places that are affected?

**Comment by jkotas** [comment]
> do you still think there are hundreds of places that are affected?

When I have introduced similar abstractions in the past, it always ended up with sizable delta. You won't know until you try it. If you believe that there are not that many, try to change the alignment to 32 bytes and get the test passing with it.

**Comment by Maoni0** [comment]
@swgillespie , let's try what Jan said. it's good learning experience regardless.

**Comment by swgillespie** [comment]
@Maoni0 ok!

**Comment by swgillespie** [comment]
I did @jkotas 's experiment of altering the alignment constant and ran into trouble on the GC side - `set_gap_size` in the GC expects that the size given to it be aligned, which in turn expects that the `gap_reloc_pair` structure be aligned. Maoni indicated that this would be thorny to fix and so I'm going to close this PR for now. I'll add a comment to the interface indicating that the alignment should not be changed.

**Pull request closed without merging**